crates/wysiwyg/src/composer_model/format.rs

```diff
@@ -51,7 +53,63 @@ where
 
     pub fn inline_code(&mut self) -> ComposerUpdate<S> {
         self.push_state_to_history();
-        self.format_or_unformat(InlineFormatType::InlineCode)
+        let format_type = InlineFormatType::InlineCode;
+        if self.action_is_reversed(format_type.action()) {
+            self.unformat(format_type)
+        } else {
+            self.add_inline_code()
+        }
+    }
+
+    /// Finds the closest structure node ancestor for each leaf node handle and groups it with other
+    /// leaves that share it as the common closest structure node ancestor. If none is found,
+    /// the root/document node is used instead.
+    pub(crate) fn group_leaves_by_closest_structure_ancestors(
+        &self,
+        leaves: Vec<&DomLocation>,
+    ) -> HashMap<DomHandle, Vec<DomLocation>> {
+        let mut structure_ancestors = HashMap::new();
+        for leaf in leaves {
+            let first_structure_ancestor =
+                self.find_structure_ancestor(&leaf.node_handle);
+            // Get the closest ancestor path or the root one (empty Vec) if there is none
+            let ancestor_handle = first_structure_ancestor
+                .map(|a| a.clone())
+                .unwrap_or(DomHandle::root());
+            let list = structure_ancestors
+                .entry(ancestor_handle)
+                .or_insert(Vec::new());
+            // Add the DomHandle of the leaf to the list of grouped handles by this ancestor
+            list.push(leaf.clone());
+        }
+        structure_ancestors
+    }
+
+    pub(crate) fn remove_and_clean_up_empty_nodes_until(
+        &mut self,
+        cur_handle: &DomHandle,
+        top_handle: &DomHandle,
+    ) {
+        self.state.dom.remove(cur_handle);
+        if cur_handle != top_handle
+            && self.state.dom.parent(cur_handle).children().is_empty()
+        {
+            self.remove_and_clean_up_empty_nodes_until(
+                &cur_handle.parent_handle(),
+                top_handle,
+            )
+        }
+    }
+
+    fn find_structure_ancestor(&self, handle: &DomHandle) -> Option<DomHandle> {
```

Comment:
Is this the same as find_struct_parent in crates/wysiwyg/src/composer_model/join_nodes.rs ?

Reply:
Ouch, I knew I had already written something like this before, but I couldn't find it. The new version seems to be a better implementation though, I'll try to de-duplicate it.

Reply:
It seems like it's also not exactly what we need here, this function actually finds *the parent of an ancestor structure node* if it exists, otherwise, that structure node (if we found a ListItem, it would return its parent List). I should probably rename it and add some docs.